docs/raw/push-notification-server.md

```diff
@@ -0,0 +1,672 @@
+
+The request MUST be sent on the [partitioned topic][./2-account.md] of the sender.
+
+The payload of the response is:
+
+```protobuf
+message PushNotificationPreferencesAcknowledgement {
+  boolean success = 1;
+  ErrorType error = 2;
+  bytes request_id = 3;
+  PushNotificationPreferences preferences = 4;
+  string access_token = 5;
+
+  enum ErrorType {
+    UNKNOWN_ERROR_TYPE = 0;
+    MALFORMED_MESSAGE = 1;
+    VERSION_MISMATCH = 2;
+    UNSUPPORTED_TOKEN_TYPE = 3;
+    INTERNAL_ERROR = 4;
+  }
+}
+```
+
+[//]: (We can ratched and use a partitioned topic here)
+
+A client SHOULD listen for an acknowledgement sent on their [partitioned topic][./2-account.md].
+
+If `success` is `true` the client has registered successfully.
+
+If `success` is `false`:
+
+- If `MALFORMED_MESSAGE` is returned, the request SHOULD not be retried without ensuring 
+  that is correctly formed.
+- If `INTERNAL_ERROR` is returned, the request MAY be retried, but the client MUST be 
+  backoff exponentially
+- If `VERSION_MISMATCH` is returned, the client SHOULD ensure that the remote version returned in `preferences` is integrated, and the version should be incremented and the request MAY be sent again.
+
+A client MAY register with multiple Push Notification Servers in order to increase availability.
+
+A client SHOULD make sure that all the notification services they registered with have the same information about their devices and tokens.
+
+If no response is returned the request SHOULD be considered failed and MAY be retried with the same server or a different one, but it MUST exponentially backoff after each trial.
+
+If the request is successful a token is returned. This SHOULD be 
+[advertised](#advertising-a-push-notification-server) as described below if no public key filtering is necessary.
+
+## Re-registering with the push notification server
+
+A client SHOULD re-register with the node if the APS or FIREBASE token changes.
+
+When re-registering a client SHOULD ensure that it has the most up-to-date
+`PushNotificationPreferences`, update the part relative to their `installation_id`
+if necessary, increment `version` and send a `PushNotificationRegister` as described above.
+
+Once re-registered, a client SHOULD advertise the changes.
+
+
+## Changing options
+
+This is handled in exactly the same way as re-registering above.
+
+## Unregistering from push notifications
+
+To unregister a client MUST send a `PushNotificationRegister` request as described
+above with `unregister` in `PushNotificationPreferences` set to `true`, or removing
+their device information.
+
+The server MUST remove all data about this user if `unregistering` is `true`,
+apart from the `hash` of the public key and the `version` of the last options,
+in order to make sure that old messages are not processed.
+
+## Advertising a push notification server
+
+Each user registered with one or more push notification servers SHOULD
```

Comment:
Just to make sure I understand correctly: Is the idea then that each user who wants to send PNs to that user would pick this up and store it locally?

Reply:
Yes, we could have a discovery for which push notification server is used by a given user, that's simple:

Every PN server listen to topic(pk-of-user) and the client can query it and the pk of the server(s) is returned, but really offers no benefits at the moment compare to just querying a mailserver, so left it out for simplicity.